File: artifacts/tulipindicators/translate_new_ablation/src/sample.rs
use rand::Rng;
use std::env;
use std::ffi::CStr;
use std::sync::MutexGuard;
use translate_without_break_down::*;
pub fn main() -> std::process::ExitCode {
    let args: Vec<String> = std::env::args().collect();
    let argc = args.len();
    let argv: Vec<&str> = args.iter().map(|s| s.as_str()).collect();

    if argc < 2 {
        println!("No indicator given.");
        println!("Example:");
        println!("\tsample ma 5");
        return std::process::ExitCode::from(1);
    }

    if argv[1] == "--version" {
        println!("TI VERSION: {}, TI BUILD: {}", "0.9.2", 1645649572);
        return std::process::ExitCode::SUCCESS;
    }

    let ti_indicators = TI_INDICATORS.lock().unwrap();
    let mut info = &ti_indicators[0];
    let mut info_idx = 0;

    if argv[1] == "--list" {
        loop {
            match info.type_ {
                1 => print!("type overlay "),
                2 => print!("type indicator "),
                3 => print!("type math "),
                4 => print!("type simple "),
                _ => print!("type unknown "),
            }

            print!("name {} ", info.name.as_ref().unwrap_or(&"".to_string()));
            print!("full_name {{{}}} ", info.full_name.as_ref().unwrap_or(&"".to_string()));

            print!("inputs {{");
            for i in 0..info.inputs as usize {
                if i > 0 {
                    print!(" ");
                }
                print!("{}", info.input_names[i].as_ref().unwrap_or(&"".to_string()));
            }
            print!("}} ");

            print!("options {{");
            for i in 0..info.options as usize {
                if i > 0 {
                    print!(" ");
                }
                print!("{{{}}}", info.option_names[i].as_ref().unwrap_or(&"".to_string()));
            }
            print!("}} ");

            print!("outputs {{");
            for i in 0..info.outputs as usize {
                if i > 0 {
                    print!(" ");
                }
                print!("{{{}}}", info.output_names[i].as_ref().unwrap_or(&"".to_string()));
            }
            print!("}}");

            println!();

            info_idx += 1;
            if info_idx >= ti_indicators.len() || ti_indicators[info_idx].name.is_none() {
                break;
            }
            info = &ti_indicators[info_idx];
        }
        return std::process::ExitCode::SUCCESS;
    }

    let indicator_name = argv[1];
    let found_info = match ti_find_indicator(Some(indicator_name)) {
        Some(found) => found,
        None => {
            println!("Couldn't find indicator: {}", argv[1]);
            return std::process::ExitCode::from(1);
        }
    };
    let info = found_info;

    // Create a vector to hold input slices
    let mut inputs: Vec<&[f64]> = Vec::new();

    let mut o = false;
    let mut h = false;
    let mut l = false;
    let mut c = false;
    let mut r = false;
    let mut r2 = false;
    let mut v = false;
    let mut a = false;

    // First collect all the data we need
    let high_data = if info.input_names.iter().any(|n| n.as_deref() == Some("high")) {
        h = true;
        Some(HIGH.lock().unwrap().to_vec())
    } else {
        None
    };

    let volume_data = if info.input_names.iter().any(|n| n.as_deref() == Some("volume")) {
        v = true;
        Some(VOLUME.lock().unwrap().to_vec())
    } else {
        None
    };

    for j in 0..info.inputs as usize {
        let empty = "".to_string();
        let input_name = info.input_names[j].as_ref().unwrap_or(&empty);
        match input_name.as_str() {
            "open" => {
                inputs.push(&OPEN);
                o = true;
            }
            "high" => {
                inputs.push(high_data.as_ref().unwrap().as_slice());
            }
            "low" => {
                inputs.push(&LOW);
                l = true;
            }
            "close" => {
                inputs.push(&CLOSE);
                c = true;
            }
            "volume" => {
                inputs.push(volume_data.as_ref().unwrap().as_slice());
            }
            "real" => {
                if !r {
                    inputs.push(&CLOSE);
                    r = true;
                } else {
                    inputs.push(&OPEN);
                    r2 = true;
                }
            }
            _ => panic!("Unknown input name"),
        }
    }

    let alts = ["acos", "asin", "atan", "cosh", "sinh", "tanh", "todeg"];
    for alt in alts.iter() {
        if *alt == info.name.as_ref().unwrap_or(&"".to_string()) {
            r = false;
            a = true;
            inputs.clear();
            for _ in 0..info.inputs as usize {
                inputs.push(&ALTERNATIVE[..]);
            }
            break;
        }
    }

    let mut options = [0.0; 10];
    for i in 0..info.options as usize {
        if argc < 3 + i {
            println!("*ERROR NOT ENOUGH OPTIONS*");
            for k in 1..argc {
                print!("{} ", argv[k]);
            }
            println!();
            return std::process::ExitCode::from(1);
        }
        options[i] = argv[2 + i].parse().unwrap_or(0.0);
    }

    let start = match &info.start {
        Some(func) => func(&options),
        None => 0,
    };

    let mut out_guard = OUT.lock().unwrap();
    // Destructure the array into separate row references
    let [row0, row1, row2] = &mut *out_guard;
    // Create mutable slices for each row
    let output0 = &mut row0[..15];
    let output1 = &mut row1[..15];
    let output2 = &mut row2[..15];
    let outputs_arr = &mut [output0, output1, output2];

    let ret = (info.indicator)(
        15,
        Some(&inputs),
        Some(&options),
        Some(outputs_arr),
    );

    if ret == 0 {
        let mut bad = false;
        print!("date        ");
        if o { print!(" open   "); }
        if h { print!(" high   "); }
        if l { print!(" low    "); }
        if c { print!(" close  "); }
        if v { print!(" volume "); }
        if r { print!(" input  "); }
        if r2 { print!(" input2  "); }
        if a { print!(" input  "); }

        for i in 0..info.outputs as usize {
            print!(" {}", info.output_names[i].as_ref().unwrap_or(&"".to_string()));
        }
        println!();

        let high_guard = HIGH.lock().unwrap();
        let volume_guard = VOLUME.lock().unwrap();

        for i in 0..15 {
            print!("{}", DATET[i]);
            if o { print!(" {:8.2}", OPEN[i]); }
            if h { print!(" {:8.2}", high_guard[i]); }
            if l { print!(" {:8.2}", LOW[i]); }
            if c { print!(" {:8.2}", CLOSE[i]); }
            if v { print!(" {:8.0}", volume_guard[i]); }
            if r { print!(" {:8.2}", CLOSE[i]); }
            if r2 { print!(" {:8.2}", OPEN[i]); }
            if a { print!(" {:8.2}", ALTERNATIVE[i]); }

            if i >= start as usize {
                for k in 0..info.outputs as usize {
                    if out_guard[k][i - start as usize].is_nan() {
                        bad = true;
                    }
                    print!(" {:8.3}", out_guard[k][i - start as usize]);
                }
            }
            println!();
        }

        if bad {
            println!("\nERROR NaN in outputs ({}).", info.name.as_ref().unwrap_or(&"".to_string()));
            return std::process::ExitCode::from(1);
        }
        return std::process::ExitCode::SUCCESS;
    } else {
        if ret == 1 {
            println!("*ERROR INVALID OPTION*");
            for i in 1..argc {
                print!("{} ", argv[i]);
            }
            println!();
        } else {
            println!("*ERROR*");
        }
        return std::process::ExitCode::from(1);
    }
}

File: artifacts/tulipindicators/decompose/src/sample.c
int helper_main_2(const ti_indicator_info * const info, unsigned int info_idx)
{
  do
  {
    if (info->type == 1)
    {
      printf("type overlay ");
    }
    else
      if (info->type == 2)
    {
      printf("type indicator ");
    }
    else
      if (info->type == 3)
    {
      printf("type math ");
    }
    else
      if (info->type == 4)
    {
      printf("type simple ");
    }
    else
      printf("type unknown ");
    printf("name %s ", info->name);
    printf("full_name {%s} ", info->full_name);
    int i;
    printf("inputs {");
    for (i = 0; i < info->inputs; i += 1)
      printf("%s%s", (i) ? (" ") : (""), info->input_names[i]);

    printf("} ");
    printf("options {");
    for (i = 0; i < info->options; i += 1)
      printf("%s{%s}", (i) ? (" ") : (""), info->option_names[i]);

    printf("} ");
    printf("outputs {");
    for (i = 0; i < info->outputs; i += 1)
      printf("%s{%s}", (i) ? (" ") : (""), info->output_names[i]);

    printf("}");
    printf("\n");
  }
  while ((++(&info[info_idx]))->name);
  return 0;
}


/*** DEPENDENCIES:
typedef struct ti_indicator_info
{
  const char *name;
  const char *full_name;
  ti_indicator_start_function start;
  ti_indicator_function indicator;
  ti_indicator_function indicator_ref;
  int type;
  int inputs;
  int options;
  int outputs;
  const char *input_names[16];
  const char *option_names[16];
  const char *output_names[16];
  ti_indicator_stream_new stream_new;
  ti_indicator_stream_run stream_run;
  ti_indicator_stream_free stream_free;
} ti_indicator_info
----------------------------
***/


// hint:  ['k_ref is a mutable refrence to int', 'bad_ref is a mutable refrence to int']
void helper_helper_main_3_1(int * const k_ref, int * const bad_ref, const ti_indicator_info * const info, int o, int h, int l, int c, int r, int r2, int v, int a, int start, int i)
{
  int k = *k_ref;
  int bad = *bad_ref;
  printf("%s", datet[i]);
  if (o)
  {
    printf(" %8.2f", open[i]);
  }
  if (h)
  {
    printf(" %8.2f", high[i]);
  }
  if (l)
  {
    printf(" %8.2f", low[i]);
  }
  if (c)
  {
    printf(" %8.2f", close[i]);
  }
  if (v)
  {
    printf(" %8.0f", volume[i]);
  }
  if (r)
  {
    printf(" %8.2f", close[i]);
  }
  if (r2)
  {
    printf(" %8.2f", open[i]);
  }
  if (a)
  {
    printf(" %8.2f", alternative[i]);
  }
  if (i >= start)
  {
    for (k = 0; k < info->outputs; k += 1)
    {
      if (out[k][i - start] != out[k][i - start])
      {
        bad = 1;
      }
      printf(" %8.3f", out[k][i - start]);
    }

  }
  printf("\n");
  *k_ref = k;
  *bad_ref = bad;
}


/*** DEPENDENCIES:
double high[] = {82.15, 81.89, 83.03, 83.3, 83.85, 83.9, 83.33, 84.3, 84.84, 85, 85.9, 86.58, 86.98, 88, 87.87}
----------------------------
double volume[] = {5653100, 6447400, 7690900, 3831400, 4455100, 3798000, 3936200, 4732000, 4841300, 3915300, 6830800, 6694100, 5293600, 7985800, 4807900}
----------------------------
double close[] = {81.59, 81.06, 82.87, 83, 83.61, 83.15, 82.84, 83.99, 84.55, 84.36, 85.53, 86.54, 86.89, 87.77, 87.29}
----------------------------
const char *datet[] = {"2005-11-01", "2005-11-02", "2005-11-03", "2005-11-04", "2005-11-07", "2005-11-08", "2005-11-09", "2005-11-10", "2005-11-11", "2005-11-14", "2005-11-15", "2005-11-16", "2005-11-17", "2005-11-18", "2005-11-21"}
----------------------------
typedef struct ti_indicator_info
{
  const char *name;
  const char *full_name;
  ti_indicator_start_function start;
  ti_indicator_function indicator;
  ti_indicator_function indicator_ref;
  int type;
  int inputs;
  int options;
  int outputs;
  const char *input_names[16];
  const char *option_names[16];
  const char *output_names[16];
  ti_indicator_stream_new stream_new;
  ti_indicator_stream_run stream_run;
  ti_indicator_stream_free stream_free;
} ti_indicator_info
----------------------------
double out[3][4000]
----------------------------
double open[] = {81.85, 81.2, 81.55, 82.91, 83.1, 83.41, 82.71, 82.7, 84.2, 84.25, 84.03, 85.45, 86.18, 88, 87.6}
----------------------------
double alternative[] = {.2, .3, .4, .3, .5, .7, .75, .9, .9, 1, 1, .2, .1, -.1, -.5}
----------------------------
double low[] = {81.29, 80.64, 81.31, 82.65, 83.07, 83.11, 82.49, 82.3, 84.15, 84.11, 84.03, 85.39, 85.76, 87.17, 87.01}
----------------------------
***/


int helper_main_3(const ti_indicator_info * const info, int o, int h, int l, int c, int r, int r2, int v, int a, int start)
{
  int i;
  int k;
  int bad = 0;
  printf("date        ");
  if (o)
  {
    printf(" open   ");
  }
  if (h)
  {
    printf(" high   ");
  }
  if (l)
  {
    printf(" low    ");
  }
  if (c)
  {
    printf(" close  ");
  }
  if (v)
  {
    printf(" volume ");
  }
  if (r)
  {
    printf(" input  ");
  }
  if (r2)
  {
    printf(" input2  ");
  }
  if (a)
  {
    printf(" input  ");
  }
  for (i = 0; i < info->outputs; i += 1)
    printf(" %s", info->output_names[i]);

  printf("\n");
  for (i = 0; i < 15; i += 1)
  {
    helper_helper_main_3_1(&k, &bad, info, o, h, l, c, r, r2, v, a, start, i);
  }

  if (bad)
  {
    printf("\nERROR NaN in outputs (%s).\n", info->name);
    return 1;
  }
  return 0;
}


/*** DEPENDENCIES:
void helper_helper_main_3_1(int * const k_ref, int * const bad_ref, const ti_indicator_info * const info, int o, int h, int l, int c, int r, int r2, int v, int a, int start, int i)
{
  int k = *k_ref;
  int bad = *bad_ref;
  printf("%s", datet[i]);
  if (o)
  {
    printf(" %8.2f", open[i]);
  }
  if (h)
  {
    printf(" %8.2f", high[i]);
  }
  if (l)
  {
    printf(" %8.2f", low[i]);
  }
  if (c)
  {
    printf(" %8.2f", close[i]);
  }
  if (v)
  {
    printf(" %8.0f", volume[i]);
  }
  if (r)
  {
    printf(" %8.2f", close[i]);
  }
  if (r2)
  {
    printf(" %8.2f", open[i]);
  }
  if (a)
  {
    printf(" %8.2f", alternative[i]);
  }
  if (i >= start)
  {
    for (k = 0; k < info->outputs; k += 1)
    {
      if (out[k][i - start] != out[k][i - start])
      {
        bad = 1;
      }
      printf(" %8.3f", out[k][i - start]);
    }

  }
  printf("\n");
  *k_ref = k;
  *bad_ref = bad;
}


----------------------------
typedef struct ti_indicator_info
{
  const char *name;
  const char *full_name;
  ti_indicator_start_function start;
  ti_indicator_function indicator;
  ti_indicator_function indicator_ref;
  int type;
  int inputs;
  int options;
  int outputs;
  const char *input_names[16];
  const char *option_names[16];
  const char *output_names[16];
  ti_indicator_stream_new stream_new;
  ti_indicator_stream_run stream_run;
  ti_indicator_stream_free stream_free;
} ti_indicator_info
----------------------------
***/


// hint:  ['o_ref is a mutable refrence to int', 'h_ref is a mutable refrence to int', 'l_ref is a mutable refrence to int', 'c_ref is a mutable refrence to int', 'r_ref is a mutable refrence to int', 'r2_ref is a mutable refrence to int', 'v_ref is a mutable refrence to int', 'j_ref is a mutable refrence to int']
void helper_main_1(int * const o_ref, int * const h_ref, int * const l_ref, int * const c_ref, int * const r_ref, int * const r2_ref, int * const v_ref, int * const j_ref, const ti_indicator_info * const info, const double *inputs[])
{
  int o = *o_ref;
  int h = *h_ref;
  int l = *l_ref;
  int c = *c_ref;
  int r = *r_ref;
  int r2 = *r2_ref;
  int v = *v_ref;
  int j = *j_ref;
  if (strcmp(info->input_names[j], "open") == 0)
  {
    inputs[j] = open;
    o = 1;
  }
  else
    if (strcmp(info->input_names[j], "high") == 0)
  {
    inputs[j] = high;
    h = 1;
  }
  else
    if (strcmp(info->input_names[j], "low") == 0)
  {
    inputs[j] = low;
    l = 1;
  }
  else
    if (strcmp(info->input_names[j], "close") == 0)
  {
    inputs[j] = close;
    c = 1;
  }
  else
    if (strcmp(info->input_names[j], "volume") == 0)
  {
    inputs[j] = volume;
    v = 1;
  }
  else
    if (strcmp(info->input_names[j], "real") == 0)
  {
    if (!r)
    {
      inputs[j] = close;
      r = 1;
    }
    else
    {
      inputs[j] = open;
      r2 = 1;
    }
  }
  else
  {
    assert(0);
  }
  ;
  *o_ref = o;
  *h_ref = h;
  *l_ref = l;
  *c_ref = c;
  *r_ref = r;
  *r2_ref = r2;
  *v_ref = v;
  *j_ref = j;
}


/*** DEPENDENCIES:
double high[] = {82.15, 81.89, 83.03, 83.3, 83.85, 83.9, 83.33, 84.3, 84.84, 85, 85.9, 86.58, 86.98, 88, 87.87}
----------------------------
double volume[] = {5653100, 6447400, 7690900, 3831400, 4455100, 3798000, 3936200, 4732000, 4841300, 3915300, 6830800, 6694100, 5293600, 7985800, 4807900}
----------------------------
double close[] = {81.59, 81.06, 82.87, 83, 83.61, 83.15, 82.84, 83.99, 84.55, 84.36, 85.53, 86.54, 86.89, 87.77, 87.29}
----------------------------
typedef struct ti_indicator_info
{
  const char *name;
  const char *full_name;
  ti_indicator_start_function start;
  ti_indicator_function indicator;
  ti_indicator_function indicator_ref;
  int type;
  int inputs;
  int options;
  int outputs;
  const char *input_names[16];
  const char *option_names[16];
  const char *output_names[16];
  ti_indicator_stream_new stream_new;
  ti_indicator_stream_run stream_run;
  ti_indicator_stream_free stream_free;
} ti_indicator_info
----------------------------
double open[] = {81.85, 81.2, 81.55, 82.91, 83.1, 83.41, 82.71, 82.7, 84.2, 84.25, 84.03, 85.45, 86.18, 88, 87.6}
----------------------------
double low[] = {81.29, 80.64, 81.31, 82.65, 83.07, 83.11, 82.49, 82.3, 84.15, 84.11, 84.03, 85.39, 85.76, 87.17, 87.01}
----------------------------
***/


int main(int argc, char **argv)
{
  const ti_indicator_info *info = ti_indicators;
  unsigned int info_idx = 0;
  if (argc < 2)
  {
    printf("No indicator given.\n");
    printf("Example:\n");
    printf("\tsample ma 5");
    return 1;
  }
  if (strcmp(argv[1], "--version") == 0)
  {
    printf("TI VERSION: %s, TI BUILD: %d\n", "0.9.2", 1746800254);
    return 0;
  }
  if (strcmp(argv[1], "--list") == 0)
  {
    helper_main_2(info, info_idx);
  }
  info_idx = ti_find_indicator(argv[1]);
  if (!(&info[info_idx]))
  {
    printf("Couldn't find indicator: %s\n", argv[1]);
    return 1;
  }
  const double *inputs[] = {0, 0, 0, 0, 0};
  double *outputs[] = {out[0], out[1], out[2], out[3], out[4]};
  int o = 0;
  int h = 0;
  int l = 0;
  int c = 0;
  int r = 0;
  int r2 = 0;
  int v = 0;
  int a = 0;
  int j;
  for (j = 0; j < info->inputs; j += 1)
  {
    helper_main_1(&o, &h, &l, &c, &r, &r2, &v, &j, info, inputs);
  }

  const char *alts[] = {"acos", "asin", "atan", "cosh", "sinh", "tanh", "todeg", 0};
  {
    const char **alt = alts;
    unsigned int alt_idx = 0;
    while (alt[alt_idx])
    {
      if (strcmp(*(&alt[alt_idx]), info->name) == 0)
      {
        r = 0;
        a = 1;
        for (j = 0; j < info->inputs; j += 1)
          inputs[j] = alternative;

        break;
      }
      alt_idx += 1;
    }

  }
  double options[10];
  {
    int i;
    for (i = 0; i < info->options; i += 1)
    {
      if (argc < (3 + i))
      {
        printf("*ERROR NOT ENOUGH OPTIONS*\n");
        for (int k = 1; k < argc; k += 1)
        {
          printf("%s ", argv[k]);
        }

        printf("\n");
        return 1;
      }
      options[i] = atof(argv[2 + i]);
    }

  }
  int start = info->start(options);
  const int ret = info->indicator(15, inputs, options, outputs);
  if (ret == 0)
  {
    helper_main_3(info, o, h, l, c, r, r2, v, a, start);
  }
  else
  {
    if (ret == 1)
    {
      printf("*ERROR INVALID OPTION*\n");
      for (int i = 1; i < argc; i += 1)
      {
        printf("%s ", argv[i]);
      }

      printf("\n");
    }
    else
    {
      printf("*ERROR*\n");
    }
    return 1;
  }
}


/*** DEPENDENCIES:
int helper_main_3(const ti_indicator_info * const info, int o, int h, int l, int c, int r, int r2, int v, int a, int start)
{
  int i;
  int k;
  int bad = 0;
  printf("date        ");
  if (o)
  {
    printf(" open   ");
  }
  if (h)
  {
    printf(" high   ");
  }
  if (l)
  {
    printf(" low    ");
  }
  if (c)
  {
    printf(" close  ");
  }
  if (v)
  {
    printf(" volume ");
  }
  if (r)
  {
    printf(" input  ");
  }
  if (r2)
  {
    printf(" input2  ");
  }
  if (a)
  {
    printf(" input  ");
  }
  for (i = 0; i < info->outputs; i += 1)
    printf(" %s", info->output_names[i]);

  printf("\n");
  for (i = 0; i < 15; i += 1)
  {
    helper_helper_main_3_1(&k, &bad, info, o, h, l, c, r, r2, v, a, start, i);
  }

  if (bad)
  {
    printf("\nERROR NaN in outputs (%s).\n", info->name);
    return 1;
  }
  return 0;
}


----------------------------
extern ti_indicator_info ti_indicators[]
----------------------------
typedef struct ti_indicator_info
{
  const char *name;
  const char *full_name;
  ti_indicator_start_function start;
  ti_indicator_function indicator;
  ti_indicator_function indicator_ref;
  int type;
  int inputs;
  int options;
  int outputs;
  const char *input_names[16];
  const char *option_names[16];
  const char *output_names[16];
  ti_indicator_stream_new stream_new;
  ti_indicator_stream_run stream_run;
  ti_indicator_stream_free stream_free;
} ti_indicator_info
----------------------------
double out[3][4000]
----------------------------
void helper_main_1(int * const o_ref, int * const h_ref, int * const l_ref, int * const c_ref, int * const r_ref, int * const r2_ref, int * const v_ref, int * const j_ref, const ti_indicator_info * const info, const double *inputs[])
{
  int o = *o_ref;
  int h = *h_ref;
  int l = *l_ref;
  int c = *c_ref;
  int r = *r_ref;
  int r2 = *r2_ref;
  int v = *v_ref;
  int j = *j_ref;
  if (strcmp(info->input_names[j], "open") == 0)
  {
    inputs[j] = open;
    o = 1;
  }
  else
    if (strcmp(info->input_names[j], "high") == 0)
  {
    inputs[j] = high;
    h = 1;
  }
  else
    if (strcmp(info->input_names[j], "low") == 0)
  {
    inputs[j] = low;
    l = 1;
  }
  else
    if (strcmp(info->input_names[j], "close") == 0)
  {
    inputs[j] = close;
    c = 1;
  }
  else
    if (strcmp(info->input_names[j], "volume") == 0)
  {
    inputs[j] = volume;
    v = 1;
  }
  else
    if (strcmp(info->input_names[j], "real") == 0)
  {
    if (!r)
    {
      inputs[j] = close;
      r = 1;
    }
    else
    {
      inputs[j] = open;
      r2 = 1;
    }
  }
  else
  {
    assert(0);
  }
  ;
  *o_ref = o;
  *h_ref = h;
  *l_ref = l;
  *c_ref = c;
  *r_ref = r;
  *r2_ref = r2;
  *v_ref = v;
  *j_ref = j;
}


----------------------------
int helper_main_2(const ti_indicator_info * const info, unsigned int info_idx)
{
  do
  {
    if (info->type == 1)
    {
      printf("type overlay ");
    }
    else
      if (info->type == 2)
    {
      printf("type indicator ");
    }
    else
      if (info->type == 3)
    {
      printf("type math ");
    }
    else
      if (info->type == 4)
    {
      printf("type simple ");
    }
    else
      printf("type unknown ");
    printf("name %s ", info->name);
    printf("full_name {%s} ", info->full_name);
    int i;
    printf("inputs {");
    for (i = 0; i < info->inputs; i += 1)
      printf("%s%s", (i) ? (" ") : (""), info->input_names[i]);

    printf("} ");
    printf("options {");
    for (i = 0; i < info->options; i += 1)
      printf("%s{%s}", (i) ? (" ") : (""), info->option_names[i]);

    printf("} ");
    printf("outputs {");
    for (i = 0; i < info->outputs; i += 1)
      printf("%s{%s}", (i) ? (" ") : (""), info->output_names[i]);

    printf("}");
    printf("\n");
  }
  while ((++(&info[info_idx]))->name);
  return 0;
}


----------------------------
const ti_indicator_info *ti_find_indicator(const char *name)
{
  int imin = 0;
  int imax = ((sizeof(ti_indicators)) / (sizeof(ti_indicator_info))) - 2;
  while (imax >= imin)
  {
    const int i = imin + ((imax - imin) / 2);
    const int c = strcmp(name, ti_indicators[i].name);
    if (c == 0)
    {
      return ti_indicators + i;
    }
    else
      if (c > 0)
    {
      imin = i + 1;
    }
    else
    {
      imax = i - 1;
    }
  }

  return 0;
}


----------------------------
double alternative[] = {.2, .3, .4, .3, .5, .7, .75, .9, .9, 1, 1, .2, .1, -.1, -.5}
----------------------------
***/


